### ISSUE TITLE
No longer compatible with Node 4

### ISSUE THREAD

**Issue by virkt25** [swagger-api/swagger-js#1315]
### Q&A (please complete the following information)
 - OS: macOS
 - Environment: Node 4.8.6
 - Method of installation: npm
 - Swagger-Client version: 3.8.1
 - Swagger/OpenAPI version: [e.g. Swagger 2.0, OpenAPI 3.0]

### Content & configuration
This is a dependency of a dependency so not sure of the implementation. 

### Describe the bug you're encountering
```
/loopback/strong-globalize/node_modules/g11n-pipeline/node_modules/swagger-client/node_modules/object-assign-deep/objectAssignDeep.js:77
function executeDeepMerge (target, _objects = [], _options = {}) {
```

`swagger-js` is no longer compatible with Node 4 due to `object-assign-deep`. Are there plans to fix it ... if not the README should be updated to reflect Node 6 and higher support 

### To reproduce...

`object-assign-deep` is Node >= 6 only. 

### Expected behavior
Expected this module to work on Node 4 / README should be updated to reflect the supported versions.

### Additional context or thoughts

Looks like `object-assign-deep` was added recently in https://github.com/swagger-api/swagger-js/pull/1309

**Comment by gzurbach** [comment]
We're experiencing the same issue. This is breaking all our builds still relying on Node 4 (over a dozen APIs). This is a major breaking change and should have been published under a new major version (i.e. `swagger-js@4.0.0`) by semver specs.

Our apps include `swagger-js@3.x` very deep in the dependency tree (we use loopback) so we cannot easily lock the version numbers.

The ideal solution would be to publish a `3.8.2` that reverts the latest changes and re-publish the latest version (`3.8.1`) as `4.0.0` while indicating the requirement for Node 6+.

**Comment by shockey** [comment]
@virkt25 + @gzurbach - sorry for the rough ride with this dependency, it looks like the module is written by hand (see https://github.com/saikojosh/Object-Assign-Deep/issues/4) and uses `const`, which breaks in Node 4.

It just so happens that we dropped Node 4 from our tests in anticipation of that version reaching End-of-Life this month (though it looks like that's [April 30th, to be exact](https://github.com/eslint/eslint/issues/10041)), so this wasn't caught before release.

I'm going to fix this, by either forking the problem module or using something else entirely, and publish a patch version for `swagger-client`. 

---

_However_, going forward from April, if you stay on Node 4 we might break your implementation without bumping the major. I'm working off of @mikeal's view of version bumping vs leaving EOL users behind (emphasis mine):

> Everything is a tradeoff. I understand that people stuck on old versions don't like this tradeoff, but you are in the minority, and **we only have the resources to support a single major line of development**, so the choice was between breaking you or opting the rest of the community out of automatic updates. This will not be the last time you are inconvenienced by changes like this in dependencies so it's a good idea for you to lock down your deps with shrinkwrap and stop updating them. Or, better yet, update to a supported version of Node.js.
> 
> For package authors: it's up to you whether or not to bump the major version, but **supporting EOL versions of Node.js is not a good practice and encourages irresponsible/insecure behavior of your users**.
> 
> https://github.com/request/request/issues/2772#issuecomment-331944074

So in short: I'm fixing this now, but _please_ upgrade your Node versions!

**Comment by gzurbach** [comment]
@shockey Thanks a lot for the quick reply. I've ben wanting to update Node in all those projects for months but my boss wants me to write new shiny toys instead. I'm sure you've heard this before 😄

What happened today will give me leverage to convince him to add this work in future sprints!

**Comment by virkt25** [comment]
Thanks for the update @shockey ! While I totally agree that EOL versions shouldn't be supported for all the reasons stated above and in the linked comment - it's a little easier said than done when enterprise customers are a bit slow to migrate and expect support. 

That said, we're working on a plan to drop Node 4 and your comment helps with that. :)

**Comment by shockey** [comment]
No worries - strictly speaking, this was on me, since v4 isn't dead quite yet.

`v3.8.2` should hit npm in a few moments, so an `npm update` should fix this issue on your end. 

Just remember that there is nothing but pain in store for you if you're still on v4 on May 1 😉 

Cheers!

**Comment by gzurbach** [comment]
Thanks for publishing `v3.8.2` it will buy us some time until we have converted all the other APIs.

FYI - I spent the last 10 hours updating some of our apps to Node 6 and I think I got it all working now. 2016 here I am, woohoo!

**Comment by virkt25** [comment]
Thanks for the fix @shockey Appreciate it and definitely gives us some time to come up with a plan to drop support for Node 4. And haha yea it's been May for a week now so you were in the right to drop support. 

@gzurbach You might get extra life if you moved on to Node 8 / 10 instead as Node 6 will be End of Life this time next year.
